_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+
+Let's also go ahead and implement the `Codec[Room]` since its very similar to `Codec[User]`:
+
+```scala
+private object SqlCommands {
+  ...
+  val roomcodec: Codec[Room] = 
+      (roomId ~ roomName).imap {
+          case i ~ n => Room(i,n)
+      }(u => (u.id, u.name))
+}
+```
+
+Next, we'll need to create its associate `Command` for the `INSERT` statement. In Skunk, `Select` statements are prepared with a `Query` while `Insert` and `Update` statements are prepared with a `Command` since it return no rows.
+
+```scala
+private object SqlCommands {
+  ...
+  val insertUser: Command[User] = 
+      sql"""
+          INSERT INTO users
+          VALUES($usercodec)
+      """
+      .command
+}
+```
+Here we prepare an insert statement, and pass the `usercodec` to `VALUES()`, then call the `command` method giving us a `Command[User]`. A `Command[User]`, means we need to pass a `User` inorder to execute the query:
+
+Let's also create `Command[Room]` since its also similar to the above:
+
+```scala
+private object SqlCommands {
+  ...
+  val insertRoom: Command[Room] =
+    sql"""
+            INSERT INTO rooms
+            VALUES($roomcodec)
+        """.command
+}
+```
+Now we can implement `createUser()`:
+
+```scala
+...
+new PostgresProtocol[F] {
+  override def createUser(name: String): F[Either[String, User]] =
+    postgres.use { session =>
+      session.prepare(insertUser).flatMap { cmd =>
+        UUIDGen.randomUUID.flatMap { id =>
+          User(id, name).flatMap {
+            case Valid(u) =>
+              cmd.execute(u).as(Right(u))
+            case Invalid(_) =>
+              Left(s"must have between 2 and 10 characters").pure[F]
+          }
+        }
+      }
+    }
+}
+```
+The `checkUser()` function takes a name of type `String` and returns an `F[Either[String, User]]`. Here we call the `use()` method on the Postgres `Resource` to access the `session` on which we call `prepare()` passing it the `insertUser` `Command`, this caches and prepares the query. 
+
+The `prepare()` method produces a `F[PreparedCommand[F, User]]`, which we flatMap on to create our `User`. We use the  `UUIDGen.randomUUID` method from cats.effect to create our `UUID` which we pass to `User` along with the `name`.
+
+If `User` creation succeeds, we call cmd.execute(u) which runs the `INSERT` query, and finally, we return a `Right(u)`. If the creation fails, we return `Left(s"must have between 2 and 10 characters").pure[F]`.
+
+Let's also implement createRoom() since it's similar:
+
+```scala
+...
+new PostgresProtocol[F] {
+  override def createRoom(name: String): F[Either[String, Room]] =
+    postgres.use { session =>
+      session.prepare(insertRoom).flatMap { cmd =>
+        UUIDGen.randomUUID.flatMap { id =>
+          Room(id, name).flatMap {
+            case Valid(r) =>
+              cmd.execute(r).as(Right(r))
+            case Invalid(_) =>
+              Left(s"must have between 2 and 10 characters").pure[F]
```

Comment:
I remember the invalid already has the message there

Reply:
Updated to pass the error message